# Informazioni dataset
Dal link https://archive.ics.uci.edu/ml/datasets/Online+Retail si estrapolano alcune informazioni per quanto riguarda gli attributi. Di seguito si elencano le cose più importanti



## Provenienza dati
Si tratta di un sito e-commerce Inglese. Vende principalmente regali per tutte le occasioni e molti dei clienti sono grossisti.

## Prezzo articolo
Nella colonna sale è presente il prezzo per singolo articolo quindi equivale a quantità=1. Il prezzo è in sterline

## Ordini cancellati

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
df = pd.read_csv("dataset/TC-dataset.csv", sep="\t",index_col=0)
df.head()

BasketID           BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0   536365  2010-12-01 08:26:00  2.55     17850.0  United Kingdom  85123A   
1   536365  2010-12-01 08:26:00  3.39     17850.0  United Kingdom   71053   
2   536365  2010-12-01 08:26:00  2.75     17850.0  United Kingdom  84406B   
3   536365  2010-12-01 08:26:00  3.39     17850.0  United Kingdom  84029G   
4   536365  2010-12-01 08:26:00  3.39     17850.0  United Kingdom  84029E   

                             ProdDescr  Qta  
0   WHITE HANGING HEART T-LIGHT HOLDER    6  
1                  WHITE METAL LANTERN    6  
2       CREAM CUPID HEARTS COAT HANGER    8  
3  KNITTED UNION FLAG HOT WATER BOTTLE    6  
4       RED WOOLLY HOTTIE WHITE HEART.    6

Secondo la descrizione del dataset quando in basket id abbiamo una C iniziale si tratta di ordini cancellati, verifichiamo quanti sono i sample cancellati

In [2]:
print(len(df[df["BasketID"].str.contains("C")]))
print(len(df[df["BasketID"].str.contains("C")].BasketID.unique()))

9084
3754


I sample totali che sono stati cancellati sono 9084, ma se guardiamo gli ordini si tratta di 3754 ordini.
Alla cancellazione degli ordini è abbinata una quantità negativa. Vediamo infatti quante sono le quantità negative del dataset e quante di queste appartendono a ordini cancellati,

In [3]:
print(len(df[df["Qta"]<=0]))
print(len(df[(df["Qta"]<=0) &(df["BasketID"].str.contains("C"))]))

9752
9084


In totale nel dataset abbiamo 9752 quantità negative, di queste 9084 appartengono a ordini cancellati.

Propongo quindi per una correttezza semantica di eliminare gli ordini cancellati

Controlliamo anche l'attributo Sale e vediamo quando è <=0

In [31]:
print(len(df[df["Sale"]<=0]))
print(len(df[(df["Sale"]<=0) &(df["BasketID"].str.contains("C"))]))

1281
0


Di conseguenza non corrisponde ad ordini cancellati

In [35]:

print(len(df[(df["Sale"]<=0) &(df["Qta"]<=0)]))

668


Praticamente tutti i casi in cui Qta<=0 o sono cancellati o hanno sale a 0

In [29]:
print(len(df[df["CustomerID"].isnull()]))
print(len(df[(df["Sale"]<=0) &(df["CustomerID"].isnull())]))

65080
1241


Quindi in corrispondenza di un sale <=0 nella maggior parte dei casi abbiamo il customerid a null

## Considerazioni finali
L'attributo Quantita<=0 indica o ordini cancellati o sale<=0 quindi:
* Propongo di eliminare tutti gli ordini con qta<=0. In questa maniera si eliminano tutti gli ordini cancellati e alcuni Sale<=0.
* I Sale<=0 rimasti li consideriamo come omaggi
